Cranelift: Add instructions for getting the current stack/frame/return pointers

fitzgen opened the pull request:
This is the initial part of https://github.com/bytecodealliance/wasmtime/issues/4535

~~TODO: I need to implement the s390x lowering for these new instructions.~~

cfallin commented on cranelift/codegen/meta/src/shared/instructions.rs:
```diff
@@ -1314,6 +1314,32 @@ pub(crate) fn define(
+    ).operands_out(vec![addr]));
+
+    ig.push(Inst::new(
+        "get_return_pointer",
```
a tiny bikeshed-color nit, but I think usually this intrinsic refers to the "return address" rather than "return pointer" (e.g., `__builtin_return_address()` in GCC/Clang); perhaps we could call it that?

cfallin commented on cranelift/codegen/src/isa/aarch64/inst/emit.rs:
```diff
@@ -1337,6 +1337,31 @@ impl MachInstEmit for Inst {
+                assert!(rm.class() == RegClass::Int);
+
+                // MOV to SP is interpreted as MOV to XZR instead. And our codegen
+                // should never MOV to XZR.
```
I would instead assert that the register is virtual; we should never have a preg-to-preg move inserted as that would either have undefined behavior (if allocatable register) or would break the ABI (move to SP or FP).

cfallin commented on cranelift/codegen/src/isa/aarch64/inst/emit.rs:
```diff
@@ -1337,6 +1337,31 @@ impl MachInstEmit for Inst {
             }
+            &Inst::MovPReg { rd, rm } => {
+                let rd = allocs.next_writable(rd);
+                let rm: Reg = rm.into();
```
Can we assert that `rm` is one of a few registers on an allow-list?

Ideally the property we want here is that it is not an allocatable register. Moving from a register managed by the regalloc has undefined behavior: we don't know what the value will be. But grabbing the `MachineEnv` and doing an assert on the lists of registers there is sort of cumbersome, so it's probably better if we limit it first to e.g. `FP` and `SP` (with a comment describing the real condition) and then add more regs to the list if we need them.

cfallin commented on cranelift/codegen/src/isa/aarch64/inst/emit.rs:
```diff
@@ -1337,6 +1337,31 @@ impl MachInstEmit for Inst {
+                // should never MOV to XZR.
+                assert!(rd.to_reg() != stack_reg());
+
+                if rm == stack_reg() {
```
Rather than duplicate the `mov` code, could we do `Inst::Mov { .. }.emit(...)`, in the same way other pseudoinsts work?

cfallin commented on cranelift/codegen/src/isa/aarch64/inst/mod.rs:
```diff
@@ -649,6 +649,11 @@ fn aarch64_get_operands<F: Fn(VReg) -> VReg>(inst: &Inst, collector: &mut Operan
         }
+        &Inst::MovPReg { rd, .. } => {
+            collector.reg_def(rd);
+            // TODO FITZGEN: should we assert that `rm` is not an allocatable
```
Ah! I see you had the same thought I did, reading a bit further :-)

I think the answer is definitely yes!

cfallin commented on cranelift/filetests/filetests/isa/x64/fp_sp_pc.clif:
```diff
@@ -0,0 +1,46 @@
+;   movq    %rsp, %rbp
+; block0:
+;   movq    %rbp, %rdi
+;   movq    8(%rdi), %rax
```
I suspect we could avoid the extra copy here by defining the return-address intrinsic to load directly from `8(%rbp)`, i.e. creating an `Amode` with `rbp` as a base register.

cfallin reviewed (commented):
A few thoughts here (and apply the same comments to x64 as for aarch64 below) but this general approach looks fine; thanks!

fitzgen commented:
@uweigand would you mind taking a look at the s390x-specific bits in here? I'm not 100% sure I got it right. I think you should be able to just look at the test expectations to quickly check here. Thanks!

cfallin commented on cranelift/codegen/src/isa/x64/encoding/rex.rs:
```diff
@@ -298,16 +298,29 @@ pub(crate) fn emit_std_enc_mem(
     prefixes.emit(sink);
 
+    let mem_e = match mem_e.clone() {
+        Amode::RbpOffset { simm32, flags } => {
```
This is to avoid providing the base reg as an operand to the operand collector, I guess?

I wonder if it might be cleaner, or less error-prone at least, to handle this case in the `Amode::get_operands` and `Amode::with_allocs` methods instead: skip providing rbp (or rsp) as an operand, and skip pulling an alloc from the AllocationConsumer if the original register was rbp (or rsp).

(I'm actually not sure; I'm going back and forth right now.)

The general concern I had with this is that we may not handle the same lowering everywhere. But actually I think everything funnels to this one lowering helper so maybe this is sufficient.

If we do stick to this approach, let's add a comment where we define the `RbpOffset` arm that it must be used for rbp (as opposed to the generic forms) because rbp cannot be given as an arg to regalloc.

fitzgen replied:
Yeah I actually like this better.

cfallin reviewed (approved):
Looks good to me! I was waffling back and forth re: the new rbp `Amode` below; on balance I think I like the approach you took better than the idea I had but I'll keep the comment for completeness. Probably enough just to add a comment describing why it's a separate Amode.

bjorn3 commented:
How are these instructions expected to behave in the face of inlining?

uweigand commented:
> @uweigand would you mind taking a look at the s390x-specific bits in here? I'm not 100% sure I got it right. I think you should be able to just look at the test expectations to quickly check here. Thanks!

Hi @fitzgen, the `get_return_address` implementation isn't correct in the general case.  It looks like you're assuming r14 holds the return address throughout execution of the function?  That isn't true, it is only guaranteed to hold the return address at the ABI boundaries, everywhere else it's just a normal allocatable general-purpose register.

For this is to be always correct, we instead have to read the return address from the stack slot where it was saved in the prologue.  However, that doesn't happen in leaf functions - unless it is forced via the `preserve_frame_pointers` flag.  Can we make that assumption for `get_return_address`?   It seems it is already made for `get_frame_pointer`.

If so, we can read the return address from the slot at "initial stack pointer + 14*8".   This could be done either via `MemArg::InitialSPOffset` (which currently isn't exposed to ISLE yet, but it should be straightforward to add a `memarg_initial_stack_off` or similar).   In the alternative, the initial stack pointer actually equals the frame pointer (backchain) value, so it could also be implemented via something like `(memarg_reg_plus_off (get_frame_pointer) 112 0 (memarg_trusted))`.

Also, I'm a bit confused about why that new `mov_preg` thing is necessary.  I'd have expected to use something like `(copy_reg (writable_reg 15))` instead, or maybe even better some new `(copy_reg (stack_reg))` - this is already used in other places to retrieve the value of a physical register.

(Or, in the alternative, even something like `(load_addr (memarg_stack_off 0))` would work to get the stack pointer.)

fitzgen commented:
> How are these instructions expected to behave in the face of inlining?

We don't have inlining so it is a bit of a hypothetical question at the moment, but these instructions are about physical frames, so I imagine we would want to ignore logical frames that don't have a corresponding physical frame (i.e. they have been inlined) and just return information about the current physical frame regardless.

fitzgen commented:
> Also, I'm a bit confused about why that new `mov_preg` thing is necessary.

It is about getting an unallocatable physical register's value into a virtual register, and avoiding the pinned vregs that we want to remove support for from `regalloc2`, and which just existed in order to get the initial migration working.

cfallin commented:
> Also, I'm a bit confused about why that new mov_preg thing is necessary. I'd have expected to use something like (copy_reg (writable_reg 15)) instead, or maybe even better some new (copy_reg (stack_reg)) - this is already used in other places to retrieve the value of a physical register.

This came out of some discussion that @fitzgen and I had: I would like to push the backends gradually away from using the "pinned vregs", or vregs that represent physical registers directly, and instead use operand constraints wherever possible. This is because handling moves-from-pinned and moves-to-pinned-vregs introduces a bunch of special cases in regalloc, and leads to worse performance (it uses a whole instruction, the move, to communicate what should be just a constraint) and sometimes bugs (e.g. bytecodealliance/regalloc2#60). Just yesterday I was helping to find the cause for a weird `mov rA, rB; mov rB, rA` sequence and it turned out to be because of moves-to-pregs before a call (in x64 in this case) where operand constraints would have led to properly elided moves.

So it's a bit forward-looking, but there is a reason! At some point later I (or someone) will go through and systematically update the remaining uses.

fitzgen commented:
Okay I pushed a couple new commits:

* switched x64 away from the new variant of `Amode` and just check the base register when getting operands for `Amode::ImmReg`, which I think is a bit nicer.
* Fixed s390x per @uweigand's feedback (thanks! would appreciate if you could give this another quick once over)

cfallin commented on cranelift/codegen/src/isa/x64/inst/args.rs:
```diff
@@ -313,14 +313,16 @@ impl Amode {
+                    collector.reg_use(*base);
+                }
             }
             Amode::ImmRegRegShift { base, index, .. } => {
```
We should assert here that base and index are not rbp or rsp I think, now that we're admitting we may use these registers in an `Amode`, just for consistency...

cfallin commented on cranelift/codegen/src/isa/x64/inst/args.rs:
```diff
@@ -313,14 +313,16 @@ impl Amode {
         match self {
             Amode::ImmReg { base, .. } => {
-                collector.reg_use(*base);
+                if *base != regs::rbp() && *base != regs::rsp() {
```
I think we need to mirror this conditional over to `with_allocs` [here](https://github.com/bytecodealliance/wasmtime/blob/c77bec4dcb33973a386d924474d95bab6b23e263/cranelift/codegen/src/isa/x64/inst/args.rs#L370) as well; otherwise we have a thing that consumes more allocs than the operands it reports. This will be kind-of-harmless today because `Amode`'s operands are always placed at the end of the inst by convention (because they're variable-length and that was easier wrt reuse-indices) but if a pseudoinst ever grows two `Amode`s or puts something after an `Amode` that could be an issue.

uweigand commented on cranelift/codegen/src/isa/s390x/inst.isle:
```diff
@@ -2473,6 +2482,20 @@
+        dst))
+
+(decl preg_r15 () PReg)
+(extern constructor preg_r15 preg_r15)
```
Just a suggestion to use something like `preg_stack` instead of r15, since this is actually not an ISA requirement - other ABIs use different registers as stack pointer.

cfallin reviewed (commented):
A few tidbits on the `RbpOffset` -> `ImmReg` change (looks good overall though!).

uweigand commented on cranelift/codegen/src/isa/s390x/inst/emit.rs:
```diff
@@ -2082,6 +2082,12 @@ impl MachInstEmit for Inst {
             }
+            &Inst::MovPReg { rd, rm } => {
+                let rm: Reg = rm.into();
+                debug_assert!([regs::gpr(14), regs::gpr(15)].contains(&rm));
```
I'd remove gpr(14), since this *is* actually allocatable.  And again preferably use stack_reg instead of hardcoding 15.

uweigand commented:
> Okay I pushed a couple new commits:
>     * Fixed s390x per @uweigand's feedback (thanks! would appreciate if you could give this another quick once over)

LGTM.  I've added a couple of inline comments, but those are really just cosmetic.

uweigand commented:
> > Okay I pushed a couple new commits:
> > * Fixed s390x per @uweigand's feedback (thanks! would appreciate if you could give this another quick once over)
> 
> LGTM. I've added a couple of inline comments, but those are really just cosmetic.

Oh, maybe one more thing comes to mind:  it would be great if the affected ISLE patterns actually *verified* that `preserve_frame_pointer` is true if they depend on that fact.   Maybe some predicate in `prelude.isle` that could be used like so:
```
(rule (lower (get_frame_pointer))
      (if (preserve_frame_pointer))
      (load64 (memarg_stack_off 0 0)))
```
Then we'd get a compile-time error instead of random run-time crashes if the assumption was violated.

uweigand commented:
> I would like to push the backends gradually away from using the "pinned vregs", or vregs that represent physical registers directly, and instead use operand constraints wherever possible.

I see, thanks for the explanation!

fitzgen commented:
> Oh, maybe one more thing comes to mind: it would be great if the affected ISLE patterns actually _verified_ that `preserve_frame_pointer` is true if they depend on that fact.

The CLIF verifier does check this condition, so by the time we lower the CLIF, we know we have valid CLIF, and this check should be unnecessary. I wouldn't mind the redundancy if it were a simple one-line debug assertion, but since it would involve new external ISLE extractors, I don't think it is quite worth it.

uweigand commented:
> > Oh, maybe one more thing comes to mind: it would be great if the affected ISLE patterns actually _verified_ that `preserve_frame_pointer` is true if they depend on that fact.
> 
> The CLIF verifier does check this condition

Ah, good!   I hadn't see that.

bjorn3 commented:
I didn't see an answer to my question in https://github.com/bytecodealliance/wasmtime/pull/4573#issuecomment-1202084202.

cfallin commented:
> I didn't see an answer to my question in [#4573 (comment)](https://github.com/bytecodealliance/wasmtime/pull/4573#issuecomment-1202084202).

@fitzgen replied to your comment in [this comment](https://github.com/bytecodealliance/wasmtime/pull/4573#issuecomment-1202975433).

bjorn3 commented:
Missed that comment. Thanks @cfallin.
